Sources/BraintreeCoreSwift/BTAnalyticsService.swift

```diff
@@ -0,0 +1,175 @@
+    }
+
+    // MARK: - Internal Methods
+
+    ///  Tracks an event.
+    ///
+    ///  Events are queued and sent in batches to the analytics service, based on the status of the app and
+    ///  the number of queued events. After exiting this method, there is no guarantee that the event has been sent.
+    /// - Parameter eventName: String representing the event name
+    func sendAnalyticsEvent(_ eventName: String) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flushIfAtThreshold()
+        }
+    }
+
+    func sendAnalyticsEvent(_ eventName: String, completion: @escaping (Error?) -> Void = { _ in }) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flush(completion)
+        }
+    }
+
+    func flush(_ completion: @escaping (Error?) -> Void = { _ in }) {
+        apiClient.fetchOrReturnRemoteConfiguration { configuration, error in
+            guard let configuration, error == nil else {
+                if let error {
+                    completion(error)
+                }
+                return
+            }
+
+            guard let analyticsURL = configuration.json?["analytics"]["url"].asURL() else {
+                completion(BTAnalyticsServiceError.missingAnalyticsURL)
+                return
+            }
+
+            if self.http == nil {
+                if let clientToken = self.apiClient.clientToken {
+                    self.http = BTHTTP(url: analyticsURL, authorizationFingerprint: clientToken.authorizationFingerprint)
+                } else if let tokenizationKey = self.apiClient.tokenizationKey {
+                    self.http = BTHTTP(url: analyticsURL, tokenizationKey: tokenizationKey)
+                } else {
+                    completion(BTAnalyticsServiceError.invalidAPIClient)
+                    return
+                }
+            }
+
+            // A special value passed in by unit tests to prevent BTHTTP from actually posting
+            if self.http?.baseURL.absoluteString == "test://do-not-send.url" {
+                completion(nil)
+                return
+            }
+
+            self.sessionsQueue.async {
+                if self.analyticsSessions.count == 0 {
+                    completion(nil)
+                    return
+                }
+
+                let willPostAnalyticsEvent = !self.analyticsSessions.keys.isEmpty
+
+                self.analyticsSessions.keys.forEach { sessionID in
+                    let postParameters = self.createAnalyticsEvent(with: sessionID)
+                    self.http?.post("/", parameters: postParameters) { body, response, error in
+                        if let error {
+                            completion(error)
+                        }
+                    }
+                }
+
+                if !willPostAnalyticsEvent {
+                    completion(nil)
+                }
+
+                completion(nil)
```

Comment:
This check isn't necessary if we'll just call the `completion(nil)` even if the check fails.

Reply:
I see in ObjC though we don't call the completion at all unless `!willPostAnalyticsEvent` passes - which is a little odd.

Reply:
Hm - weird, maybe the tests are timing out and false-positive passing? I would need to look at which specific test you mean. 

But what I meant was if we always want to call the completion(nil), we can just remove the 
```
                if !willPostAnalyticsEvent {
                    completion(nil)
                }
```

entirely.

Reply:
Yep, I knew what you meant, was just kind of a odd discovery. The tests aren't false positive passing, I checked that as well, but it seems like there is some other _weird magic_ (probably not actually magic just weirdness)